File: java/src/main/java/problems1_100/ContainerWithMostWater.java
package problems1_100;

enum ContainerWithMostWater {
    DEFAULT {
        @Override
        int maxArea(int[] height) {
            int start = 0;
            int end = height.length - 1;
            int maxArea = Integer.MIN_VALUE;
            while (start < end) {
                int wall = Math.min(height[start], height[end]);
                maxArea = Math.max(maxArea, (end - start) * wall);
                if (height[start] < height[end]) {
                    start++;
                } else {
                    end--;
                }
            }
            return maxArea;
        }
    };

    abstract int maxArea(int[] height);
}

File: kotlin/src/main/kotlin/problems1_100/ContainerWithMostWater.kt
package problems1_100

import kotlin.math.max
import kotlin.math.min

enum class ContainerWithMostWater {
    DEFAULT {
        override fun maxArea(height: IntArray): Int {
            var start = 0
            var end = height.lastIndex
            var maxArea = Int.MIN_VALUE
            while (start < end) {
                val wall = min(height[start], height[end])
                maxArea = max(maxArea, ((end - start) * wall))
                if (height[start] < height[end]) {
                    start++
                } else {
                    end--
                }
            }
            return maxArea
        }
    },
    ;

    abstract fun maxArea(height: IntArray): Int
}
